Aggressive slicer v2

polgreen opened the pull request:
This is a new version of #1587

It is dependent on:
#2381
#2369

The aggressive slicer by default removes function bodies of any functions not on the shortest path on the call graph between the start function and the property. If no property is specified, it will preseve the shortest path for each property. It is parameterisable:

--call-depth " preserves all functions within function calls of the shorets path " --preserve-function forces the aggressive slicer to preserve function 
--preserve-function containing force the aggressive slicer to preserve all functions with names containing 
--preserve-all-direct-paths force the aggressive slicer to preserve all functions on direct paths to the property. This option must be used with a specified property.

The aggressive slicer is designed to be used in conjunction with CBMC inbuilt features to block functions and to replace function bodies with approximations. 

The results are not sound, it may produce spurious traces, due to over-approximating the removed function bodies, and it may miss traces due to the removed function bodies writing to global variables, or if "preserve-all-direct-paths" is not used. However, it is designed for use with code bases that are far too big to otherwise use cbmc on.

The parameterisation is intended to be used as a means of providing engineer feedback.

polgreen committed:
aggressive slicer regression tests

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 5dcb64a85869a8f5bcd227e21b946d695dbb512b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76958248

smowton commented on regression/goto-instrument/aggressive_slicer1/main.c:
```diff
@@ -0,0 +1,30 @@
+  	D();
+}
+
+void B() 
```
So my guess is we'll fail to reach D because only the B -> C -> D path is actually feasible, but it isn't the shortest path to D? If so note that as a comment in either the test or the .desc file

polgreen replied:
I can't find any other regression tests with documentation; can you point me to an example?

tautschnig replied:
`regression/contracts/function_apply_01/` or indeed all of them in `regression/contracts/` have info, in those cases both in the C source and the test.desc file.

smowton replied:
Example syntax:

```
CORE
PrimaryArg
otherarg1 otherarg2
regex that must match
another that must match
--
regex that must not match
another that must not match
--
comments go here, below the second -- delimiter
comments continue on subsequent lines.
```

Alternatively comment in the `.c` file for the regression test

smowton commented on regression/goto-instrument/aggressive_slicer5/main.c:
```diff
@@ -0,0 +1,29 @@
+  __CPROVER_assert(0,"");
+}
+
+void C()
```
I think `aggressive_slicer5` is missing a .desc file

smowton commented on regression/goto-instrument/aggressive_slicer5/main.c:
```diff
@@ -0,0 +1,29 @@
+void D()
```
Similarly for the rest of these tests -- briefly comment stating the intent of the test, either in the source or the desc file

smowton commented on regression/goto-instrument/aggressive_slicer6/main.c:
```diff
@@ -0,0 +1,29 @@
+void D()
```
Consider adding a test case with more than one property, so specifying a property makes a difference?

smowton commented on src/goto-instrument/aggressive_slicer.cpp:
```diff
@@ -0,0 +1,112 @@
+ * aggressive_slicer.cpp
+ *
+ *  Created on: 18 Jun 2018
+ *      Author: polgreen
```
Author lines should give an email

smowton commented on src/goto-instrument/aggressive_slicer.cpp:
```diff
@@ -0,0 +1,112 @@
+#include <goto-programs/show_properties.h>
+#include <util/message.h>
+
+void aggressive_slicert::slice_for_function(irep_idt destination_function)
```
Rename, as no slicing is being done here. `note_functions_to_keep` perhaps?

smowton commented on src/goto-instrument/aggressive_slicer.cpp:
```diff
@@ -0,0 +1,112 @@
+    {
+      for(const auto &ins : fct.second.body.instructions)
+        if(ins.is_assert())
+          slice_for_function(ins.function);
```
Is `ins.function` always equal to `fct.first`, or does inlining mean it could be otherwise? Either way suggest keeping a set of functions already done, so you don't go pointlessly rebuilding shortest paths for functions you've already marked.

smowton commented on src/goto-instrument/aggressive_slicer.cpp:
```diff
@@ -0,0 +1,112 @@
+  }
+}
+
+void aggressive_slicert::find_functions_that_contain_name_snippet()
```
`keep_functions_that_contain_name_snippet`?

smowton commented on src/goto-instrument/aggressive_slicer.cpp:
```diff
@@ -0,0 +1,112 @@
+{
+  messaget message(message_handler);
+
+  functions_to_keep.insert(CPROVER_PREFIX "initialize");
```
Use https://github.com/diffblue/cbmc/blob/develop/src/linking/static_lifetime_init.h#L24

smowton commented on src/goto-instrument/aggressive_slicer.cpp:
```diff
@@ -0,0 +1,112 @@
+
+  // Necessary to compute cone of influence
+  if(preserve_all_direct_paths)
+    disconnect_unreachable_functions(call_graph, start_function);
```
Why build then prune, rather than using https://github.com/diffblue/cbmc/blob/develop/src/analyses/call_graph.h#L31 ?

smowton commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1433,6 +1434,46 @@ void goto_instrument_parse_optionst::instrument_goto_program()
+    status() << "Performing an aggressive slice" << eom;
+    aggressive_slicert aggressive_slicer(goto_model, get_message_handler());
+
+    if(cmdline.isset("call-depth"))
```
Suggest prefixing these with `aggressive-slicer-` to make clear they are specific to a particular transform

smowton commented on src/analyses/call_graph_helpers.cpp:
```diff
@@ -102,6 +102,15 @@ std::set<irep_idt> get_functions_reachable_within_n_steps(
+  const irep_idt &function)
+{
+  std::size_t start_idx = *(graph.get_node_index(function));
+  if(start_idx != optionalt<std::size_t>())
```
If it was Empty like this the deref on the line above would already have faulted. This should probably be
```
if(auto start_idx = graph.get_node_index(function))
  graph.disconnect_unreachable(*start_idx);
```

(This line compiled because `optionalt<std::size_t>` has an implicit conversion from `std::size_t`, so the current test is equivalent to `if(optionalt<std::size_t>(start_idx) != optionalt<std::size_t>())`, which is always true)

polgreen replied:
There don't actually seem to be any checks that optional arguments have been returned elsewhere in this file, e.g., 
` graph.get_reachable(*(graph.get_node_index(function)), forwards);` is used in `get_connected_functions`. 

Do I need the check? What happens in the above case if the function isn't found and therefore no argument is returned?

tautschnig replied:
The `*` operator would cause a failing assertion. If you want to do a check, don't apply `operator*` and test `.has_value()` of the result before using (i.e., dereferencing).

polgreen replied:
So, if I want to throw an error anyway, if no argument is returned, I may as well just do what is done in the get_reachable function?

smowton replied:
If you're happy with an assertion failure then yes simply remove the `if(start_idx != optionalt<std::size_t>())`. Since it's using old-fashioned `assert` the result will be an assert failure on a debug build, or continuing with undefined behaviour on a release build.

smowton commented on src/analyses/call_graph_helpers.cpp:
```diff
@@ -101,3 +101,25 @@ std::set<irep_idt> get_functions_reachable_within_n_steps(
+  std::size_t start_node = *graph.get_node_index(src);
+  std::size_t dest_node = *graph.get_node_index(dest);
+
+  PRECONDITION(
```
As the other instance, dereferencing an Empty `optionalt` throws an exception, so make sure you check for emptiness (usually with `.has_value()` or the implicit boolean conversion (e.g. `if(graph.get_node_index(src)) { // now safe to use *graph.get_node_index(src) }`) rather than comparison with `optionalt<std::size_t>()` before using `operator*`

smowton commented on src/analyses/call_graph_helpers.cpp:
```diff
@@ -101,3 +101,25 @@ std::set<irep_idt> get_functions_reachable_within_n_steps(
     graph, start_functions, n, forwards);
 }
+
+std::set<irep_idt> get_shortest_function_path(
```
Surprised this is set -- I'd expect a path to be ordered

smowton commented on unit/analyses/call_graph.cpp:
```diff
@@ -223,6 +223,48 @@ SCENARIO("call_graph",
+        std::size_t depth = 1;
+        std::set<irep_idt> reachable = get_functions_reachable_within_n_steps(
+          exported, function_name, depth, true);
+        REQUIRE(reachable.size() == 2);
```
Could use `REQUIRE(reachable == std::set<irep_idt> {"A", "B"});` and similar

polgreen replied:
This check is the same code as is used in several other unit tests in this file, is there a reason mine should be different? Unit tests that use this are:
get_callees: https://github.com/diffblue/cbmc/blob/develop/unit/analyses/call_graph.cpp#L229-L231
get_callers: https://github.com/diffblue/cbmc/blob/develop/unit/analyses/call_graph.cpp#L236-L238
get_reaching_functions: https://github.com/diffblue/cbmc/blob/develop/unit/analyses/call_graph.cpp#L243-L249
etc.

smowton replied:
Not really, looks like I wrote those too, just a suggestion of a maybe cleaner way of making the same assertion

smowton commented on src/util/graph.h:
```diff
@@ -584,6 +602,92 @@ std::vector<typename N::node_indext> grapht<N>::get_reachable(
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const std::vector<node_indext> &src,
+  std::size_t &limit,
```
Pass this by value

smowton commented on src/util/graph.h:
```diff
@@ -584,6 +602,92 @@ std::vector<typename N::node_indext> grapht<N>::get_reachable(
+{
+  PRECONDITION(nodes.size() < std::numeric_limits<std::size_t>::max());
+
+  if(limit <= 0)
```
`std::size_t` is unsigned, so `== 0`

smowton commented on src/util/graph.h:
```diff
@@ -584,6 +602,92 @@ std::vector<typename N::node_indext> grapht<N>::get_reachable(
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const std::vector<node_indext> &src,
+  std::size_t &limit,
+  std::vector<bool> &visited,
```
How about making this a `std::set<node_indext> &`, then while the membership test would be a little more expensive than a `vector<bool>` you could use it as the result, rather than copying the result after each recursive call. The function would look something like

```
depth_limited_search(node_indext from, std::size_t steps, std::set<node_indext> &result)
{
  for(const auto o : nodes[from].out)
    if(result.insert(o.first).second)
      depth_limited_search(o.first, steps - 1, result);
}
```

smowton commented on src/util/graph.h:
```diff
@@ -584,6 +602,92 @@ std::vector<typename N::node_indext> grapht<N>::get_reachable(
+/// should be performed.
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const std::vector<node_indext> &src,
```
This would be much simpler if it worked over single sources rather lists-of-sources

polgreen replied:
Yes it would. I wrote the function to work with a vector of sources to be consistent with `grapht::get_reachable`, which has a function for a vector of sources and one for a single source. I don't know when you would want to do reachability from multiple sources, but given that is an option, it seems nicely symmetric to also have an option to do depth-limited reachability from multiple sources.

smowton replied:
Sure, but the multi-source one can just be a wrapper, like
```
std::set<irep_idt> depth_limited_search(
  const std::vector<node_indext> &srcs, size_t depth)
{
  std::set<irep_idt> result;
  for(const node_indext src : srcs)
    depth_limited_search(src, result, depth);
  return result;
}
```

tautschnig replied:
As discussed with @polgreen out-of-band: This doesn't quite work, you need to maintain a collection of nodes visited along some path, or else capture in some other way what your exploration frontier is. For example, a graph with nodes {a,b,c,d} and edges a->b->c, a->c->d must yield "d" for depth 2 even though "c" is already seen on one path.

smowton replied:
Argh, true, sorry. Would breadth-first search do the trick?

smowton replied:
(In case you've trashed the old version of this but want to get back to it -- `git reflog` lets you see where a branch used to point to and recover the old "lost" commits)

tautschnig replied:
Re BFS: I think you get in trouble whenever you use an implementation that uses recursion, because you need to communicate the depth of when you have seen a node across invocations.

polgreen replied:
I used BFS last time I implemented this,  and, if I recall correctly, @smowton reviewed it and said that I should switch it to DFS in order to use more efficient containers. 

(The way to do it with BFS is to mark the end of each frontier with some specific node index, I think I used numeric_limits::max)

smowton replied:
I wasn't gunning for DFS in particular, I was trying to eliminate the copying of the results or frontier vector at each recursive call. However I've clearly made far more mess than is justified here, so I'll butt out.

polgreen replied:
The BFS implementation didn't involve recursion and copying, just a for loop. But it didn't use vectors, it used a stack. 

Either way, I'd prefer it if we just stuck with one version now so that I can redo all my changes on top of that. it's easier for me to revert to the DFS version, since the BFS version was based on the old call graph class so I would pretty much have to rewrite it.

smowton replied:
Oh I see, I thought you were referring to the review a few days ago, not a really old one

smowton commented on unit/analyses/disconnect_unreachable_nodes_in_graph.cpp:
```diff
@@ -0,0 +1,223 @@
+#include <goto-programs/goto_convert_functions.h>
+
+static symbolt
+create_void_function_symbol(const irep_idt &name, const codet &code)
```
If these are being duplicated, suggest putting them in `unit/call_graph_test_utils.h` or something like that

smowton commented on src/util/graph.h:
```diff
@@ -461,6 +464,43 @@ void grapht<N>::visit_reachable(node_indext src)
+    else if(i == reachable[reachable_idx])
+      reachable_idx++;
+    else if(i > reachable[reachable_idx])
+      throw "error disconnecting unreachable nodes";
```
This is an invariant rather than a throw, as it's a programming mistake rather than a user error / bad input

smowton reviewed (commented):
Just realised I've also commented on the dependent PRs here. Hopefully this is useful anyhow.

Major things:
* Comments on tests indicating their intent
* `depth_limited_search` could be much simpler

Otherwise minor comments

polgreen committed:
Aggressive slicer

polgreen commented:
changes done

tautschnig commented:
I'll merge #2381 once CI passes so that this PR is reduced to a unique set of commits once rebased.

polgreen committed:
remove mistake in grapht

Function declaration exists but there is no body. Removing function declaration

polgreen committed:
call graph helper functions

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: f35cc0c49889ca3c57d61fba09273e9ec0e49315).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77436678

polgreen commented:
this is now rebased

tautschnig commented on regression/goto-instrument/aggressive_slicer1/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+^warning: ignoring
```
Nit pick: github reminds us that there is no newline at the end of the file.

tautschnig commented on regression/goto-instrument/aggressive_slicer2/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+^warning: ignoring
```
Missing newline

tautschnig commented on regression/goto-instrument/aggressive_slicer3/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+^warning: ignoring
```
Missing newline

tautschnig commented on regression/goto-instrument/aggressive_slicer4/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+^warning: ignoring
```
Missing newline

tautschnig commented on regression/goto-instrument/aggressive_slicer5/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+^warning: ignoring
```
Missing newline

tautschnig commented on regression/goto-instrument/aggressive_slicer6/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+^warning: ignoring
```
Missing newline

tautschnig commented on regression/goto-instrument/aggressive_slicer7/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+^warning: ignoring
```
Missing newline

polgreen commented:
Are there any further changes needed on this? I've incorporated all the changes requested above

polgreen commented:
I've made all the changes requested on this and rebased it; are there any further changes required?

tautschnig reviewed (commented):
I seems I never found the time to complete this review: some nitpicks below.

tautschnig commented on src/goto-instrument/aggressive_slicer.cpp:
```diff
@@ -0,0 +1,115 @@
+#include <goto-programs/show_properties.h>
+#include <util/message.h>
+
+void aggressive_slicert::note_functions_to_keep(irep_idt destination_function)
```
Pass by (const) reference.

polgreen commented:
Changes made

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: de97624301f0718d1d113c455bfec391af330179).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77942419

polgreen committed:
Apply clang format to includes in goto_instrument_parse_options

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: c456098278209e66d22fe24820d59202ad4329ce).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78169164

polgreen commented:
The tests pass on this, and the changes requested have all been made, is there anything else that needs doing? 

It appears to have requested additional reviewers because I applied clang-format to goto_instrument_parse_options.cpp. I only applied the clang-format patch that was provided by the CI. I've left it in a separate commit because it is pretty unrelated to anything in this PR, but it is also necessary to get this PR to pass all the CI.

polgreen commented:
Are there any additional changes required to this PR? If not, is there a reason it hasn't been merged?

martin-cs commented on src/analyses/call_graph_helpers.cpp:
```diff
@@ -74,10 +74,10 @@ std::set<irep_idt> get_reaching_functions(
 {
-  std::vector<std::size_t> start_indices;
   std::set<irep_idt> result;
+  std::vector<std::size_t> start_indices;
```
This looks a little like an inconsistency when updating.  Where the two PRs that this depended on changed and this undoes some of the changes?

polgreen replied:
One of the PRs that this depends on ended up in a mess, because I did a major requested change, then did some minor changes on top of that, and then had to reverse the major change. And so there are some bits and pieces that have ended up in this PR that shouldn't be, like passing std::size_t by reference. Those changes are deliberate. 

Switching the order of start_indices and result is unintentional though, I've reversed that now

martin-cs replied:
Thanks.  I appreciate that some of this is a bit pedantic and my thanks for your patience.  I just want to avoid getting into a fix / undo because PR raced with fix / fix loop.

peterschrammel replied:
I think @martin-cs' comment should be addressed.

martin-cs commented on src/util/graph.h:
```diff
@@ -255,11 +255,6 @@ class grapht
-    const node_indext &src,
-    std::size_t &limit,
-    bool forwards) const;
-
```
Why is this a mistake?  Is there a body that needs removal as well?

polgreen replied:
No, the mistake is that the body was already removed

martin-cs replied:
OK.  In future it would be nice to have that in a commit message.

peterschrammel replied:
It would be good to amend the commit message according to @martin-cs' comment.

martin-cs commented on src/goto-instrument/aggressive_slicer.h:
```diff
@@ -0,0 +1,100 @@
+  /// slicer, i.e., shortest path between two functions, or all direct paths.
+  /// Inserts functions to preserve into the functions_to_keep set
+  /// \param destination_function name of destination function for slice
+  void note_functions_to_keep(const irep_idt &destination_function);
```
Wow!  Actual useful documentation!

martin-cs commented on src/goto-instrument/goto_instrument_parse_options.h:
```diff
@@ -90,6 +90,11 @@ Author: Daniel Kroening, kroening@kroening.com
+  "(aggressive-slice-call-depth):" \
+  "(aggressive-slice-preserve-function):" \
+  "(aggressive-slice-preserve-functions-containing):" \
+  "(aggressive-slice-preserve-all-direct-paths)" \
```
The prevalent trend for these seems to be as an OPT_ macro, along with the coresponding help macro.

peterschrammel replied:
I think this comment should be addressed.

martin-cs reviewed (approved):
A few things it would be good for you to look at but I'm not going to request this be held / re-reviewed once you've had a look at them.

peterschrammel commented on regression/goto-instrument/aggressive_slicer5/test.desc:
```diff
@@ -0,0 +1,8 @@
+CORE
+main.c
```
The file aggressive_slicer5/test.binary must not be committed.

peterschrammel commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1457,6 +1458,46 @@ void goto_instrument_parse_optionst::instrument_goto_program()
+      aggressive_slicer.call_depth =
+        safe_string2unsigned(cmdline.get_value("aggressive-slice-call-depth"));
+
+    if(cmdline.isset("aggressive-slice-preserve-function"))
```
I don't quite understand the use of having this option plus `aggressive-slice-preserve-functions-containing`. Why not allow a regex to specify what should be preserved?

polgreen replied:
It could be done with a regex, I didn't need the full capabilities of regex though and found this easier to work with.

peterschrammel replied:
I'm saying that there are standard ways of specifying name filtering (wild cards, Regex). 'contains name snippet' is not among these and may confuse users. Using regexes would even save you a couple of lines in the implementation.

peterschrammel commented on src/goto-instrument/aggressive_slicer.h:
```diff
@@ -0,0 +1,100 @@
+      goto_model(_goto_model),
+      message_handler(_msg)
+  {
+    start_function = goto_model.goto_functions.entry_point();
```
Not sure why this needs to be stored, and if necessary, make it `const`.

polgreen replied:
it's being used multiple times, so seemed better to store it than to call `goto_model.goto_functions.entry_point()` multiple times

peterschrammel commented on src/goto-instrument/aggressive_slicer.h:
```diff
@@ -0,0 +1,100 @@
+      functions_to_keep.insert(f);
+  };
+
+  std::list<std::string> user_specified_properties;
```
Why are these public?

polgreen replied:
these are written to in goto_instrument_parse_options.cpp

peterschrammel commented on src/goto-instrument/aggressive_slicer.h:
```diff
@@ -0,0 +1,100 @@
+      message_handler(_msg)
+  {
+    start_function = goto_model.goto_functions.entry_point();
+    call_depth = 0;
```
prefer assignment at declaration or in initializer

peterschrammel commented on src/goto-instrument/aggressive_slicer.h:
```diff
@@ -0,0 +1,100 @@
+  bool preserve_all_direct_paths;
+
+private:
+  irep_idt start_function;
```
remove or make `const`

peterschrammel commented on src/goto-instrument/aggressive_slicer.h:
```diff
@@ -0,0 +1,100 @@
+  void get_all_functions_containing_properties();
+};
+
+#endif /* CPROVER_GOTO_INSTRUMENT_AGGRESSIVE_SLICER_H  */
```
Use // comments

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 7be6f999e54fca419bd41a1aa797f87c023b3ef6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79781452

polgreen committed:
move aggressive slicer options into macro

polgreen committed:
make comment // not /*

polgreen commented:
I've made the changes except the regex, which would be an extension of the functionality of this PR. Pending the CI, please could this be merged?

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 8c7dc1789f65a0f076bee60d811b3419531144cd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79824378
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##
